### PR TITLE
gameEvent: Implement Throw Card Event

### LINKED ISSUE #44: Add a game event handler for the "THROW_CARD" event.
Please refer to [this file](https://github.com/shivansh-bhatnagar18/multiplayer-uno/blob/master/backend/uno-game-engine/gameEvents.ts)

# Description
The game engine has an event based architecture: Each game event is handled by a dedicated handler. 
We can implement the `THROW_CARD` event now, since the card creation logic is complete.

# Tasks
- Since different cards behave differently, the handler should consider all such cases. For example, it should reverse the turn flow if a reverse card is thrown etc.
- It should also check whether it is possible to throw a card or not, and return success or error accordingly.
- The handler logic should be verified through tests using appropriate testcases.

# Notes
Please refer to the global types related to this.

### PR DESCRIPTION
**Pull request by criticic**
## Description
adds the throw card handler

Fixes #44 

## Motivation and Context
[Explain the motivation behind these changes and provide any relevant context.]

## How to Test
[Describe the steps to test the changes made in this pull request.]

## Related Issues
[If applicable, mention any related issues or pull requests.]

## Checklist
- [ ] I have tested these changes locally.
- [ ] I have reviewed the code and ensured it follows the project's coding guidelines.
- [ ] I have updated the documentation, if necessary.
- [ ] I have assigned reviewers to this pull request.

## Screenshots (if applicable)
[If your changes include any visual updates, provide screenshots here.]

### CODE REVIEW

File: backend/uno-game-engine/engine.ts [hunk 1]
```diff
@@ -47,11 +47,19 @@ export class GameEngine {
...
-    drawCardFromDeck(player: Player) {
+    drawCardFromDeck(player: Player, numCards = 1) {
         //todo: Handle the case when the deck is empty and we have to move the thrown cards back to the deck
-        this.players
-            .find((p: Player) => p.id === player.id)
-            .cards.push(this.cardDeck.pop());
+        // this.players
+        //     .find((p: Player) => p.id === player.id)
+        //     .cards.push(this.cardDeck.pop());
+        const currentPlayer = this.players.find((p) => p.id === player.id);
```
[the remark this answers is not in the document]
Comment (criticic) [reply backend/uno-game-engine/engine.ts:55]: #1 is another blocker for testing rn

[the remark this answers is not in the document]
Comment (kuv2707) [reply backend/uno-game-engine/engine.ts:55]: Done with it.

File: backend/uno-game-engine/events/throwCard.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    if (!player) {
+        return { type: 'ERROR', message: 'Player not found' };
+    }
+    const card = findCard(player, event.data.card.id);
+    if (!card) {
+        return { type: 'ERROR', message: 'Card not found' };
+    }
+
+    // check if the card can be thrown
+    if (game.lastThrownCard && !canThrowCard(game.lastThrownCard, card)) {
```
Comment (kuv2707) [inline backend/uno-game-engine/events/throwCard.ts:32]: move all the checks to `canThrowCard`

File: backend/uno-game-engine/events/throwCard.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    // check if the card is present in the player's hand
+    const player = findPlayer(game, event.playerId);
+    if (!player) {
+        return { type: 'ERROR', message: 'Player not found' };
+    }
+    const card = findCard(player, event.data.card.id);
+    if (!card) {
+        return { type: 'ERROR', message: 'Card not found' };
+    }
```
Comment (kuv2707) [inline backend/uno-game-engine/events/throwCard.ts:29]: We should also check whether the player actually possesses said card

**Review by kuv2707 — COMMENTED** [review]
Looks great!
Posted few comments.

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -26,11 +26,17 @@ format of the message is:
 
 ```json
 {
-	"type": "DRAW_CARD",
-	"player": "player1",
+	"type": "THROW_CARD",
+	"playerId": "1",
 	"data": {
-		"cardID": "red-5"
+		"card": {
```
[the remark this answers is not in the document]
Comment (criticic) [reply ARCHITECTURE.md:37]: I'll do it in a moment, rest of the things, i feel i have done

**Comment by kuv2707** [comment]
@criticic Please prioritize this PR over any new work as this will help further work on the engine.

**Comment by kuv2707** [comment]
@criticic Are you working on it or should I close it?

**Comment by criticic** [comment]
I’ll send a updated PR by tonight

On 11 Jun 2024, at 11:14 AM, Kislay ***@***.***> wrote:



@criticic<https://github.com/criticic> Are you working on it or should I close it?

File: ARCHITECTURE.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (criticic) [reply ARCHITECTURE.md:37]: Is there much to do in this PR, apart from the documentation change, I guess the api handler will have a get card from card id function

File: backend/src/types.d.ts [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // We declare those types which are used throughout the application here.
 // For types that are used only in one file, we can declare them in that file itself.
 
-type CardType = 'number' | 'special' | 'wild';
+type CardType = 'number' | 'special';
```
[the remark this answers is not in the document]
Comment (kuv2707) [reply backend/src/types.d.ts:4]: I wasn't aware of it either. But the rules we are following (link in ARCHITECTURE.md) mention it and it seems fun to implement.
Even if that weren't the case, it is a good idea to model the types after the actual game objects/properties.

File: ARCHITECTURE.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kuv2707) [reply ARCHITECTURE.md]: You should address all feedback

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -31,13 +31,12 @@ format of the message is:
...
 ```json
 {
-	"type": "DRAW_CARD",
-	"player": "player1",
-	"data": {
-		"cardID": "red-5"
-	}
+	"type": "THROW_CARD",
+	"playerId": "1",
+	"cardId": "card-number-red-5",
```
Comment (kuv2707) [inline ARCHITECTURE.md:36]: We'd want to go with the data object here.

File: backend/src/uno-game-engine/events/throwCard.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    return { type: 'SUCCESS', message: 'Can throw card' };
+}
+
+export function throwCard(game: GameEngine, event: GameEvent): EventResult {
+    const player = findPlayer(game, event.playerId);
+    if (!player) {
+        return { type: 'ERROR', message: 'Player not found' };
+    }
+
+    const card = findCard(player, event.data.card.id);
```
Comment (kuv2707) [inline backend/src/uno-game-engine/events/throwCard.ts:47]: It seems unnecessary to send the whole card object from the client when we can use id. I had earlier mentioned that we must setup a function to retrieve cards by id. Seems like you missed it.

File: backend/src/uno-game-engine/events/throwCard.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            break;
+        case 'draw2':
+            {
+                const nextPlayer = game.players[game.currentPlayerIndex];
+                game.drawCardFromDeck(nextPlayer, 2);
+            }
+            break;
+        case 'draw4':
+            {
+                const currentPlayer = game.players[game.currentPlayerIndex];
```
Comment (kuv2707) [inline backend/src/uno-game-engine/events/throwCard.ts:88]: since `nextPlayer` is not called before this, `currentPlayerIndex` will still refer to the player who just threw the card.

File: backend/tests/engine.test.ts
Comment (kuv2707) [inline backend/tests/engine.test.ts:1]: These provide foundational tests for the game engine, which is good to have.

**Review by kuv2707 — APPROVED** [review]
The logic needs some minor reworking. For the record, the `THROW_CARD` event object will have `cardId` as the only attribute in its `data`.
Also:
- We should shift the getting-card-by-id function to deck.ts.
- As mentioned earlier, it is very important to have tests for the throw-card logic because of subtle corner cases.

I think we can defer the tests for now, given that we are a little behind the schedule, but please be motivated to follow up on this with the tests. (you can open a PR for that even without any issue).

Despite some minor issues, I think we can merge it as it is good enough for the time being. I'll make some of the amendments in subsequent commits.

Some general suggestions:
- Respond to **all** feedback, even with an emoji reaction. This lets the maintainer know that the particular feedback has been addressed.
- Tag the maintainer if you want another review.

**Comment by kuv2707** [comment]
Merged, thanks @criticic

**Comment by criticic** [comment]
Just pushed a commit, did that end up being merged?

On 13 Jun 2024, at 12:11 AM, Kislay ***@***.***> wrote:



Merged #50<https://github.com/shivansh-bhatnagar18/multiplayer-uno/pull/50> into master.

**Comment by criticic** [comment]
hey @kuv2707 please see the latest branch: https://github.com/criticic/csoc-multiplayer-uno/tree/feat/event-throw-card
